server/src/server/app.py

```diff
@@ -22,49 +28,105 @@
-
-    # Load character configuration
-    print("Loading character configuration")
-    character = loadCharacters(os.getenv("CHARACTER_FILE", "rolypoly.json"))[0]
-    personality = process_character_config(character)
-    
-    # Combine base instructions with character config
-    full_instructions = BASE_INSTRUCTIONS.format(
-        character_instructions=personality,
-        character_name=character["name"],
-        adjectives=", ".join(character.get("adjectives", [])),
-        topics=", ".join(character.get("topics", []))
-    )
-    print("Full instructions:", full_instructions)
-    for tool in TOOLS:
-        print(tool.name)
-
-    
-    agent = OpenAIVoiceReactAgent(
-        model="gpt-4o-realtime-preview",
-        tools=TOOLS,
-        instructions=full_instructions,
-        voice="verse" #"alloy", "ash", "ballad", "coral", "echo", "sage", "shimmer", and "verse"
-    )
 
-    await agent.aconnect(browser_receive_stream, websocket.send_text)
+async def websocket_endpoint(websocket: WebSocket):
+    global active_connections
+
+    # Connection limiting
+    if active_connections >= MAX_CONNECTIONS_PER_INSTANCE:
+        await websocket.accept()
+        await websocket.send_text(
+            json.dumps(
+                {
+                    "type": "error",
+                    "message": "Server at capacity. Please try again later.",
+                }
+            )
+        )
+        await websocket.close()
+        return
+
+    active_connections += 1
+    try:
+        await websocket.accept()
+
+        browser_receive_stream = websocket_stream(websocket)
+
+        # Load character configuration
+        print("Loading character configuration")
+        character = loadCharacters(os.getenv("CHARACTER_FILE", "rolypoly.json"))[0]
+        personality = process_character_config(character)
+
+        # Combine base instructions with character config
+        full_instructions = BASE_INSTRUCTIONS.format(
+            character_instructions=personality,
+            character_name=character["name"],
+            adjectives=", ".join(character.get("adjectives", [])),
+            topics=", ".join(character.get("topics", [])),
+        )
+        print("Full instructions:", full_instructions)
+        for tool in TOOLS:
+            print(tool.name)
+
+        agent = OpenAIVoiceReactAgent(
+            model="gpt-4o-realtime-preview",
+            tools=TOOLS,
+            instructions=full_instructions,
+            voice="verse",  # "alloy", "ash", "ballad", "coral", "echo", "sage", "shimmer", and "verse"
+        )
+
+        await agent.aconnect(browser_receive_stream, websocket.send_text)
+    finally:
+        active_connections -= 1
 
 
 async def homepage(request):
     with open("server/src/server/static/index.html") as f:
         html = f.read()
         return HTMLResponse(html)
 
-routes = [Route("/", homepage), WebSocketRoute("/ws", websocket_endpoint)]
+
+async def health_check(request):
```

Comment:
`/health` endpoint added for fly.io monitoring.  When fly.io detects that the machine health is deteriorating, it scales up new instances (configured in the fly.toml).